pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -166,36 +170,71 @@ func generateMetadataStringMatcher(keys []string, v *metadata.StringMatcher) *me
 	return &metadata.MetadataMatcher{
 		Filter: authn.AuthnFilterName,
 		Path:   paths,
-		Value: &metadata.MetadataMatcher_Value{
-			MatchPattern: &metadata.MetadataMatcher_Value_StringMatch{
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_StringMatch{
 				StringMatch: v,
 			},
 		},
 	}
 }
 
-// createDynamicMetadataMatcher creates a MetadataMatcher for the given key, value pair.
-func createDynamicMetadataMatcher(k, v string) *metadata.MetadataMatcher {
+func generateMetadataListMatcher(keys []string, v string) *metadata.MetadataMatcher {
+	listMatcher := &metadata.ListMatcher{
+		MatchPattern: &metadata.ListMatcher_OneOf{
+			OneOf: &metadata.ValueMatcher{
+				MatchPattern: &metadata.ValueMatcher_StringMatch{
+					StringMatch: createStringMatcher(v, false),
+				},
+			},
+		},
+	}
+
+	paths := make([]*metadata.MetadataMatcher_PathSegment, 0)
+	for _, k := range keys {
+		paths = append(paths, &metadata.MetadataMatcher_PathSegment{
+			Segment: &metadata.MetadataMatcher_PathSegment_Key{Key: k},
+		})
+	}
+
+	return &metadata.MetadataMatcher{
+		Filter: authn.AuthnFilterName,
+		Path:   paths,
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_ListMatch{
+				ListMatch: listMatcher,
+			},
+		},
+	}
+}
+
+// Generate keys for metadata paths
+func generateMetaKeys(k string) ([]string, error) {
 	var keys []string
-	forceRegexPattern := false
 
 	if k == attrSrcNamespace {
 		// Proxy doesn't have attrSrcNamespace directly, but the information is encoded in attrSrcPrincipal
 		// with format: cluster.local/ns/{NAMESPACE}/sa/{SERVICE-ACCOUNT}, so we change the key to
-		// attrSrcPrincipal and change the value to a regular expression to match the namespace part.
+		// attrSrcPrincipal.
 		keys = []string{attrSrcPrincipal}
-		v = fmt.Sprintf(`*/ns/%s/*`, v)
-		forceRegexPattern = true
 	} else if strings.HasPrefix(k, attrRequestClaims) {
 		claim, err := extractNameInBrackets(strings.TrimPrefix(k, attrRequestClaims))
 		if err != nil {
-			return nil
+			return nil, err
 		}
 		keys = []string{attrRequestClaims, claim}
+	} else if strings.HasPrefix(k, authDerivedClaims) {
```

Comment:
derived claims should not be user facing.

Reply:
Done.